TokenBucket.php

```diff
@@ -0,0 +1,146 @@
+ * also specifies the maximum number of tokens the bucket can hold.
+ *
+ * The current number of tokens is time sensative so it is not stored as a
+ * member of the class, instead you will have to call getTokens() to calculate
+ * and retrieve the current number of tokens.
+ *
+ * Last consume can possibly be changed by another concurrent request, and
+ * therefore is also not stored in the Object's state and always retrieved from
+ * storage when needed.
+ */
+class TokenBucket {
+   // String to identify this bucket uniquely in storage.
+   private $identifier;
+   // TokenRate is when your parents try to ration your arcade spending.
+   private $rate;
+   // Where the token is to be stored.
+   private $storage;
+
+   /**
+    * @param $identifier string to uniqely store the bucket.
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate) {
+      if (!is_string($key)) {
+         throw new InvalidArgumentException("identifier must be a string");
+      }
+
+      $this->key = $key;
+      $this->backend = $backend;
+      $this->rate = $rate;
+   }
+
+   /**
+    * Tries to consume a token, if a token can't be consumed, then a DateTime
+    * is returned for when the next token will be available.
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *               index 1 being a timestamp of when there will be
+    *               enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_int($amount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      $storedBucket = $this->backend->get($this->key);
+      if ($storedBucket === Backend::MISS) {
+         $storedBucket = new StoredBucket($this->rate->tokens, microtime(true));
+      }
+      $tokens = $storedBucket->getTokens();
+      $lastConsume = $storedBucket->getLastConsume();
+
+      $updatedTokens = $tokens - $amount;
+      $now = microtime(true);
+      $newBucket = new StoredBucket($updatedTokens, $now);
+      if ($updatedTokens < 0) {
+         return [false, $this->readyTime($amount, $newBucket)];
+      }
+
+      $this->storeBucket($newBucket);
+      return [true, $now];
+   }
+
+   /**
+    * @return $tokens int the total number of tokens in the bucket currently.
+    */
+   public function getTokens() {
+      $storedBucket = $this->backend->get($this->key);
+      if ($storedBucket === Backend::MISS) {
+         return $this->rate->tokens;
+      }
+
+      $currentTokens = $storedBucket->getTokens();
+      $lastConsume = $storedBucket->getLastConsume();
+      $updatedTokens = $this->updateTokens($currentTokens, $lastConsume);
+      return $updatedTokens;
+   }
+
+   /**
+    * Updates tokens to the amount that it should be after restoring the tokens
+    * that have been regenerated from the last consume to now.
+    */
+   private function updateTokens($tokens, $lastConsume) {
```

Comment:
Since this function has no side-effects, I think `getAvailableTokens($storedBucket)` might be a better name and interface.